server/src/main/java/org/elasticsearch/action/bulk/Retry2.java

```diff
@@ -0,0 +1,419 @@
+        } else {
+            logger.trace("Rejecting a bulk request because the queue is full. Queue size is {}", readyToLoadQueue.size());
+            listener.onFailure(
+                new EsRejectedExecutionException("Could not load bulk request, bulk request queue at capacity [" + queueCapacity + "]")
+            );
+        }
+    }
+
+    /**
+     * Retries the bulkRequestForRetry if retriesRemaining is greater than 0, otherwise notifies the listener of failure
+     * @param bulkRequestForRetry The bulk request for retry. This should only include the items that have not previously succeeded
+     * @param responsesAccumulator An accumulator for all BulkItemResponses for the original bulkRequest across all retries
+     * @param consumer
+     * @param listener The listener to be notified of success or failure on this retry or subsequent retries
+     * @param retriesRemaining The number of times remaining that this BulkRequest can be retried
+     */
+    private void retry(
+        BulkRequest bulkRequestForRetry,
+        List<BulkItemResponse> responsesAccumulator,
+        BiConsumer<BulkRequest, ActionListener<BulkResponse>> consumer,
+        ActionListener<BulkResponse> listener,
+        int retriesRemaining
+    ) {
+        /*
+         * Here we calculate when this request will next be up for retry (in clock time) and put it on retryQueue. We use nanonTime rather
+         * than currentTimeInMillis because nanoTime will not change if the system clock is updated. We do not actually care about
+         * nanosecond-level resolution.
+         */
+        if (retriesRemaining > 0) {
+            addToQueue(bulkRequestForRetry, responsesAccumulator, consumer, listener, retriesRemaining - 1);
+        } else {
+            listener.onFailure(
+                new EsRejectedExecutionException(
+                    "Could not queue bulk request for retry because the backoff policy does not allow any more retries"
+                )
+            );
+        }
+    }
+
+    /**
+     * This method servers two purposes: (1) It promotes bulk requests from the retryQueue to the readyToLoadQueue when their time to run
+     * has come and (2) It calls the consumer bulk requests that are on the readyToLoadQueue.
+     */
+    private void flush() {
+        while (isClosing == false || System.nanoTime() < closingTime) {
+            boolean allowedToMakeRequest;
+            long timeRemaining = isClosing ? closingTime - System.nanoTime() : 0;
+            if (isClosing && timeRemaining > 0) {
+                try {
+                    logger.trace(
+                        "Waiting up to {} for a semaphore because the server is closing",
+                        new TimeValue(timeRemaining, TimeUnit.NANOSECONDS)
+                    );
+                    allowedToMakeRequest = requestsInFlightSemaphore.tryAcquire(timeRemaining, TimeUnit.NANOSECONDS);
```

Comment:
This seems strange to me. Requests consume resources (i.e. heap) whether they are in flight or just waiting on the queue, so I think it's important to bound this number. Conversely there shouldn't really be a need to limit the number of requests that are actually in flight to Elasticsearch like this. It seems unlikely we'd actually overwhelm Elasticsearch with these requests, and even if we did we should just get back an `EsRejectedExecutionException` or `CircuitBreakingException` to indicate a need to retry.

Also do we really need to block a thread like this? I know it's our own flush thread so it's kind of ok but it still seems strange. This is effectively reimplementing a scheduler and thread pool queue, but I don't see a good reason for having to do this.

Reply:
```Conversely there shouldn't really be a need to limit the number of requests that are actually in flight to Elasticsearch like this.```
We had been limiting this previously, so I carried the limit forward here. But I guess it's a good point -- if we're changing to a model where we're throwing `EsRejectedExecutionException` if the queue is full it makes sense to just do that here instead, and not use the semaphore.

Reply:
Let me see if I'm following where you're going with this:
We keep track of how many bytes are in flight total for BulkProcessor (globally), and reject any requests that come in if that number is > X. So we don't need a queue or a semaphore or a thread to manage it. When we get a BulkRequest (including a retry), we check `if(inflightBytes + newBytes > X) throw new EsRejectedExecutionException`. Otherwise we add `newBytes` to `inflightBytes` and submit the request. When its listener is notified we subtract its number of bytes.

Reply:
Exactly, yes.

Reply:
Oh wait, not quite:

> When we get a BulkRequest (including a retry)

I think not for retries, they're kind of an internal thing so we should already be tracking the bytes held by the request we're retrying. We'd only release things once the request is complete (either it's fully successful or we decide to stop retrying for some other reason).

Reply:
Yeah that makes sense.